File: Algorithm-CPP/55_기차운행.cpp
#include <iostream>
#include <stack>
#include <vector>
using namespace std;

int main() {
    vector<char> res;
    stack<int> stack;
    int n, cur = 1, index = 0;
    cin >> n;

    int arr[n];
    for(int i = 0; i < n; i++) {
        cin >> arr[i];
    }

    while(index < n) {
        if(arr[index] == cur) {
            res.push_back('P');
            res.push_back('O');
            index++; cur++;

            while(!stack.empty()) {
                if(stack.top() == cur) {
                    stack.pop();
                    res.push_back('O');
                    cur++;
                } else {
                    break;
                }
            }
        } else {
            stack.push(arr[index]);
            res.push_back('P');
            index++;
        }
    }

    if(!stack.empty()) {
        cout << "impossible" << endl;
        return 0;
    }

    for(int i = 0; i < res.size(); i++) {
        cout << res[i];
    }
    cout << endl;

    return 0;    
}

File: Algorithm-Swift/Algorithm-Swift/Inflearn/55_기차운행.swift
//
//  55_기차운행.swift
//  Algorithm-Swift
//
//  Created by SHIN YOON AH on 2023/08/27.
//

import Foundation

func solution55() {
    let n = Int(readLine()!)!
    let arr = readLine()!.split(separator: " ").map { Int($0)! }
    var stack: [Int] = []
    var cur = 1, index = 0, res = ""

    while index < n {
        if arr[index] == cur {
            res += "PO"
            index += 1
            cur += 1

            while !stack.isEmpty {
                if stack.last! == cur {
                    stack.removeLast()
                    res += "O"
                    cur += 1
                } else { break }
            }
        } else {
            stack.append(arr[index])
            index += 1
            res += "P"
        }
    }

    if !stack.isEmpty {
        print("impossible")
        return
    }

    print(res)
}
